### PR TITLE
15fcos: remember when Ignition ran and print on console

### PR DESCRIPTION
**Pull request by jlebon**
Some users sometimes may not realize that they're using a pre-booted
version of a CoreOS image. This makes things confusing because they then
don't understand why the Ignition config wasn't applied.

There's no way to consistently detect this, but at least we can print an
informational message about (1) when Ignition ran, and (2) how many
boots ago that was.

This enhances the Ignition issue we already write for whether a
user config was provided rather than creating a separate one.

Related: https://bugzilla.redhat.com/show_bug.cgi?id=1977949
Related: https://github.com/coreos/ignition/issues/1214

### CODE REVIEW

**Comment by jlebon** [comment]
TBH, I'm not a big fan of this myself because in 99.9% of cases, it's just additional noise on the console. I think we can tackle subsets of it more smartly (e.g. https://github.com/coreos/ignition/issues/1214 for the forced Ignition side, or verifying that if it's `ConditionFirstBoot=true`, then we must be in an Ignition boot; that would catch a tool which scrubs `/etc/machine-id`, or checking SSH host keys, etc...). But I'm not sure how to tackle this in the general case where a user simply boots an image, writes back, then boots it back thinking it's the first boot ([this](https://github.com/coreos/fedora-coreos-config/blob/3568dc2e7f691ed6abf8a37242a616b9eaf9ba5b/overlay.d/15fcos/usr/lib/systemd/system/coreos-check-ignition-config.service#L5-L9) is related).

**Comment by darkmuggle** [comment]
I'm not sure that having a boot counter is the best proxy. Reading the bug, the root question is about the provisioning status. I suspect reporting "provisioned" based on success of https://github.com/coreos/ignition/blob/main/systemd/ignition-firstboot-complete.service would give the user more detailed information.

**Comment by jlebon** [comment]
> I'm not sure that having a boot counter is the best proxy. Reading the bug, the root question is about the provisioning status. I suspect reporting "provisioned" based on success of [https://github.com/coreos/ignition/blob/main/systemd/ignition-firstboot-complete.service](https://github.com/coreos/ignition/blob/main/systemd/ignition-firstboot-complete.service?rgh-link-date=2021-07-06T17%3A28%3A24Z) would give the user more detailed information.

My idea was "if we're going to always take up a line of input on the console, then we might as well add something dynamic that's somewhat useful". If boot is > 1, then it's already been provisioned (at least at the RHCOS/FCOS level). But yeah agreed it might not be clear enough.

Hmm OK, how about instead we print the *date* the machine was provisioned? So e.g.

> Provisioned: Tue 06 Jul 2021 05:41:30 PM UTC

and if it's the current boot:

> Provisioned: Tue 06 Jul 2021 05:41:30 PM UTC (current boot)

And we have `ignition-firstboot-complete.service` write out a stamp file somewhere with the date it ran? (Since journal logs eventually get GC'ed so we can't just always query the journal.)

**Comment by darkmuggle** [comment]
> > I'm not sure that having a boot counter is the best proxy. Reading the bug, the root question is about the provisioning staus. I suspect reporting "provisioned" based on success of [https://github.com/coreos/ignition/blob/main/systemd/ignition-firstboot-
> Hmm OK, how about instead we print the _date_ the machine was provisioned? So e.g.
> 
> > Provisioned: Tue 06 Jul 2021 05:41:30 PM UTC
> 
> and if it's the current boot:
> 
> > Provisioned: Tue 06 Jul 2021 05:41:30 PM UTC (current boot)
> 
> And we have `ignition-firstboot-complete.service` write out a stamp file somewhere with the date it ran? (Since journal logs eventually get GC'ed so we can't just always query the journal.)

I'm a fan. Also the only thing I'd suggest is `Ignition Provisioning Completed: <DATE> <current boot>` to show that Ignition was done, while other provisioning tools and systems (perhaps the MCO) could could still be doing its thing.

**Comment by jlebon** [comment]
OK let's do https://github.com/coreos/fedora-coreos-config/pull/1087 first and then I'll build on top of that.

**Comment by jlebon** [comment]
OK updated this now! Instead of emitting another issue, I enhanced the existing one we already emitted for whether an Ignition config was provided. Sample output:

```
Fedora CoreOS 34.20210708.dev.0
Kernel 5.12.13-300.fc34.x86_64 on an x86_64 (ttyS0)

SSH host key: SHA256:uAN0XBl05ln929Jnb8OzjsAwEtT96h9hgqC3PpBfz9k (ECDSA)
SSH host key: SHA256:jud2D6IEirH+aljDmwXltr/MKkOKQi8PbCrKHFYM8co (ED25519)
SSH host key: SHA256:MViYRbGL8MPSfe9CG2UTq4UKhRYDfnehzU6uLH5wVo4 (RSA)
No SSH authorized keys provided by Ignition or Afterburn
Ignition: ran on Thu Jul  8 19:16:43 UTC 2021 (this boot) with Ignition config
localhost login:
```

```
Fedora CoreOS 34.20210708.dev.0
Kernel 5.12.13-300.fc34.x86_64 on an x86_64 (ttyS0)

SSH host key: SHA256:TFaFUyCPrNYaMfRPfD9RFAIMrR/OmqbhwCM10Yg5y8I (ECDSA)
SSH host key: SHA256:yy4lUsMa0F2Uv/hdDBArNlUHrlaPhH2ejPjOKsoEr08 (ED25519)
SSH host key: SHA256:eC1ED8njg+nG6gtPqDDvz3O+bxWDjQoLCDL8E3O99lQ (RSA)
ens3: 10.0.2.15 fec0::8498:52db:5dd:314a
No SSH authorized keys provided by Ignition or Afterburn
Ignition: ran on Thu Jul  8 19:19:13 UTC 2021 (at least 2 boots ago) without Ignition config
ibm-p8-kvm-03-guest-02 login:
```

**Comment by jlebon** [comment]
OK, couldn't help tinkering with this a bit more. I think it reads better now.

```
Fedora CoreOS 34.20210708.dev.0
Kernel 5.12.13-300.fc34.x86_64 on an x86_64 (ttyS0)

SSH host key: SHA256:EmNciosfHowZY8deNnicM4W21gyg/+ipiOVHlJltQcU (ECDSA)
SSH host key: SHA256:PgUwx4HZicZx/pF+MbXIenZ2ZsvWj61NXjtsKFY8U1A (ED25519)
SSH host key: SHA256:O8v+j12TLbrAvmHlD73o09zqZsoeqXLpLw2bGUnlp2I (RSA)
No SSH authorized keys provided by Ignition or Afterburn
Ignition: config applied on Thu Jul  8 19:56:47 UTC 2021 (this boot)
localhost login:
```

```
Fedora CoreOS 34.20210708.dev.0
Kernel 5.12.13-300.fc34.x86_64 on an x86_64 (ttyS0)

SSH host key: SHA256:8qOqZl8iD4XRPyjn7zVpCCNPzErq1ubSTgTPQqFIXuw (ECDSA)
SSH host key: SHA256:ZUxkcQR84e3kbM3d3rx/y2+7xOpqCfawrFo3KI2w8jc (ED25519)
SSH host key: SHA256:z4z65X6tjsgwvx0QMr7RL1MipTiiDAffikPbX/yeX6k (RSA)
ens3: 10.0.2.15 fec0::dcf5:5a8e:1d50:aae9
No SSH authorized keys provided by Ignition or Afterburn
Ignition: ran on Thu Jul  8 19:54:21 UTC 2021 (at least 2 boots ago) without a config
ibm-p8-kvm-03-guest-02 login:
```

**Comment by dustymabe** [comment]
hmm.. bikeshed commence.. I almost prefer to leave the existing statements and provide a new one.

```
Ignition: executed on 2021/07/01 15:09:01 UTC (1 boot ago)
Ignition: user provided config was applied
Ignition: wrote ssh authorized keys file for user: core
```

```
Ignition: executed on 2021/07/01 15:09:01 UTC (this boot)
Ignition: no config provided by user
No ssh authorized keys provided by Ignition or Afterburn
```

also prefer the more succinct date format: `date +"%Y/%m/%d %H:%M:%S %Z"`

**Comment by jlebon** [comment]
> hmm.. bikeshed commence.. I almost prefer to leave the existing statements and provide a new one.

Open to that. Will wait some more in case there are other suggestions to discuss.

**Comment by jlebon** [comment]
Updated this now!

```
Fedora CoreOS 34.20210716.dev.1
Kernel 5.12.15-300.fc34.x86_64 on an x86_64 (ttyS0)

SSH host key: SHA256:1+T/QzmdgedFmpHhu+7R1oauY7W8HKq6Z14q4ETA5zM (ED25519)
SSH host key: SHA256:A8NbdOkEQ7H5gDnthxahvpcpBpN1zci/qz3q3Xi2Mzw (ECDSA)
SSH host key: SHA256:mTsNm5OklEqrzDtjqANhlQaumCkPd6Fg4vTbD5WqN4k (RSA)
ens7: 10.0.2.15 fe80::501a:4032:507a:8b9c
No SSH authorized keys provided by Ignition or Afterburn
Ignition: ran on 2021/07/16 15:51:17 UTC (this boot)
Ignition: user-provided config was applied
cosa-devsh login:
```

```
Fedora CoreOS 34.20210716.dev.1
Kernel 5.12.15-300.fc34.x86_64 on an x86_64 (ttyS0)

SSH host key: SHA256:Vy/DGIIIGy+1y0f3NLszR/oR8GkrbZuk4HPR9WZUQwU (ECDSA)
SSH host key: SHA256:DIa+tzP90mqWWIldal5qdrPRYQ9jE9K/ufNVVFGV9lY (ED25519)
SSH host key: SHA256:dawfRNpXssvjCncD4CqH3UsE6U9OTn02TRI3PfFbYr8 (RSA)
ens3: 10.0.2.15 fec0::7050:7317:e0fb:5c4
No SSH authorized keys provided by Ignition or Afterburn
Ignition: ran on 2021/07/16 15:55:05 UTC (at least 2 boots ago)
Ignition: no config provided by user
ibm-p8-kvm-03-guest-02 login:
```

File: overlay.d/15fcos/usr/libexec/coreos-check-ignition-config [hunk 1]
```diff
@@ -1,26 +1,42 @@
...
-    echo "Ignition: user-provided config was applied" > /etc/issue.d/30_ignition_config_info.issue
+is_firstboot=0
+if [ -e "${IGNITION_FIRSTBOOT_INFO}" ]; then
+    is_firstboot=1
+    mkdir -p "$(dirname "${IGNITION_INFO}")"
+    mv "${IGNITION_FIRSTBOOT_INFO}" "${IGNITION_INFO}"
+fi
+
+mkdir -p /run/issue.d
+touch /run/issue.d/coreos-ignition-provisioning.issue
```
Comment (dustymabe) [inline overlay.d/15fcos/usr/libexec/coreos-check-ignition-config]: previously the filename we used had a number in front (for sorting). Is there good reason to leave that out here?

File: overlay.d/15fcos/usr/libexec/coreos-check-ignition-config [hunk 2]
```diff
@@ -1,26 +1,42 @@
...
+RESET='\033[0m' # reset
 
-if [[ $output -gt  0 ]];then
-    echo "Ignition: user-provided config was applied" > /etc/issue.d/30_ignition_config_info.issue
+is_firstboot=0
+if [ -e "${IGNITION_FIRSTBOOT_INFO}" ]; then
+    is_firstboot=1
+    mkdir -p "$(dirname "${IGNITION_INFO}")"
+    mv "${IGNITION_FIRSTBOOT_INFO}" "${IGNITION_INFO}"
+fi
```
Comment (dustymabe) [inline overlay.d/15fcos/usr/libexec/coreos-check-ignition-config:16]: I feel like this could use a comment.

File: overlay.d/15fcos/usr/libexec/coreos-check-ignition-config [hunk 3]
```diff
@@ -1,26 +1,42 @@
...
+mkdir -p /run/issue.d
+touch /run/issue.d/coreos-ignition-provisioning.issue
+
+d=$(date --date "@$(jq .date "${IGNITION_INFO}")" +"%Y/%m/%d %H:%M:%S %Z")
+if [ "${is_firstboot}" == 1 ]; then
+    echo "Ignition: ran on ${d} (this boot)" \
+        > /run/issue.d/coreos-ignition-provisioning.issue
+else
+    nreboots=$(($(journalctl --list-boots | wc -l) - 1))
+    [ "${nreboots}" -eq 1 ] && boot=boot || boot=boots
```
Comment (dustymabe) [inline overlay.d/15fcos/usr/libexec/coreos-check-ignition-config]:
```suggestion
    [ "${nreboots}" -eq 1 ] && boot='boot' || boot='boots'
```

This is easier to read for me if it's explicit that the right hand side is a string (I know this is bash, but we switch between languages a lot and this will just help).

File: overlay.d/15fcos/usr/libexec/coreos-check-ignition-config [hunk 4]
```diff
@@ -1,26 +1,42 @@
...
+        >> /run/issue.d/coreos-ignition-provisioning.issue
 else
-    echo -e "${warn}Ignition: no config provided by user${nc}" > /etc/issue.d/30_ignition_config_info.issue
+    echo -e "${WARN}Ignition: no config provided by user${RESET}" \
+        >> /run/issue.d/coreos-ignition-provisioning.issue
 fi
 
-# Ask all running agetty instances to reload and update their
-# displayed prompts in case this script was run before agetty.
-/usr/sbin/agetty --reload
```
Comment (dustymabe) [inline overlay.d/15fcos/usr/libexec/coreos-check-ignition-config:26]: previously we were running `agetty --reload`. No longer needed? Maybe we should make that a separate commit with explanation?

**Review by dustymabe — COMMENTED** [review]
Looking good. Some final comments

**Review by dustymabe — APPROVED** [review]
LGTM

File: overlay.d/15fcos/usr/libexec/coreos-check-ignition-config (hunk 4 above)
Comment (travier) [reply overlay.d/15fcos/usr/libexec/coreos-check-ignition-config:26]: I have the same question here. Was this resolved?

Comment (dustymabe) [reply overlay.d/15fcos/usr/libexec/coreos-check-ignition-config:26]: yes it was. It wasn't obvious, but he added a second commit to the PR that explained why it was OK to remove it.
